Add more tests and jsdocs for saltcorn-admin-models & Add ability to limit file types to upload

AleksandrSokolov opened the pull request:
1. Add more tests and jsdocs for saltcorn-admin-models
2. Add ability to limit file types to upload

glutamate commented:
@AleksandrSokolov it would be better to add the file filter to the upload fileview here:  

https://github.com/saltcorn/saltcorn/blob/v0.8.0-beta.0/packages/saltcorn-data/base-plugin/fileviews.js#L101-L128

Add your field (https://github.com/saltcorn/saltcorn/pull/1472/files#diff-b67356cdcfedae3f65d51bff81d7350911b0be04245d24021b5b9feac7c983afR597-R602) here instead: https://github.com/saltcorn/saltcorn/blob/v0.8.0-beta.0/packages/saltcorn-data/base-plugin/fileviews.js#L113

and your implementation of the accepts attr to the input here: https://github.com/saltcorn/saltcorn/blob/v0.8.0-beta.0/packages/saltcorn-data/base-plugin/fileviews.js#L124

The admin should be able to upload any kind of file, but restricting the user is very useful (e.g. only images)

AleksandrSokolov commented:
@glutamate please review. I reworked file filtering. Added accept filter to file field. So for every field column possible to define own filter. Also default filter applied if user does not define filter for particular field.

glutamate commented on packages/server/routes/fields.js:
```diff
@@ -310,6 +312,15 @@ const fieldFlow = (req) =>
                     "Deleting a row will also delete the file referenced by this field"
                   ),
                 },
+                {
```
adding file filter here is clever. great idea

AleksandrSokolov replied:
Will fixed soon.

glutamate commented on packages/saltcorn-data/models/table.ts:
```diff
@@ -405,7 +405,7 @@ class Table implements AbstractTable {
         for (const row of rows) {
           // run triggers on delete
-          await trigger.run!(row);
+          trigger.run!(row);
```
need await here

AleksandrSokolov replied:
Ok, will fixed soon.

AleksandrSokolov replied:
fixed

glutamate commented on packages/saltcorn-data/models/table.ts:
```diff
@@ -910,18 +910,27 @@ class Table implements AbstractTable {
     try {
-      const s = await getLines(filePath, 500);
-      rows = await csvtojson().fromString(s); // todo agrument type unknown
+      let lines_limit = state.getConfig("csv_lines_limit");
```
The 500 in the previous code was not the limit to the CSV files to be passed. Instead this is the number of lines that are used to determine the types in the created table. The file is read twice, but on the first run we don't need to traverse all of it

AleksandrSokolov replied:
Yes. This is clear, Idea to add parameter - that 500 rows can be not relevant in some cases to determine types in created tables. So I tried to give more flexibility to admin for tune this process. I think that needs to change name of "csv_lines_limit"  to better name. But no idea of better name))

AleksandrSokolov replied:
I got errors when load big csv files. When 500 rows were not enough to determine types.

AleksandrSokolov replied:
@glutamate, name changed to "csv_types_detection_rows". Please review

glutamate commented on packages/saltcorn-data/models/config.ts:
```diff
@@ -593,6 +593,28 @@ const configTypes: ConfigTypes = {
+    blurb: "Specifies a default filter for what file types the user can pick from the file input dialog box. Example is `.doc, text/csv,audio/*,video/*,image/*`",
+  },
+  /** @type {object} */
+  csv_lines_limit: {
```
remove this, see comment on table

AleksandrSokolov replied:
@glutamate, Please see my comments in discussion about table

glutamate commented:
Thank you, looks good!
